### PR TITLE
feat!: update experimentalists `grid_pool`, `random_pool`, `random_sample` to use the new `State` mechanism

### LINKED ISSUE #9: grid pooler and random pooler expect different arguments
The grid pooler expects a list of IVs (or at least variables) as input, while the random pooler expects a list of discrete values. It might be worth making this consistent.

### PR DESCRIPTION
**Pull request by hollandjg**
Updates:
- **Breaking Change**, _Refactor_: remove structural `pooler` vs. `sampler` split in `autora.experimentalist`
- **Breaking Change**, _Feature_: Update standard way of setting up experimentalists so that every function should be wrapped by the end-user, reducing load on contributors and hopefully making reuse easier
- **Breaking Change**, _Refactor_: Gather standard State and wrapper functions in a single submodule to allow `from autora.state.standard import StandardState, Delta, on_state, state_fn_from_estimator`
- **Breaking Change**, _Feature_: convert existing example experimentalists `grid_pool`, `random_pool`, and `random_sample` to use the `State + Delta` mechanism with the standard field names from `StandardState` 
- CI: Update Pre-Commit Hooks

Resolutions:
- resolves #9 
- requires #32 
- resolves https://github.com/AutoResearch/autora/issues/231
- resolves https://github.com/AutoResearch/autora/issues/462

See also: 
- https://github.com/AutoResearch/autora-experimentalist-sampler-model-disagreement/pull/10

## Discussion – moving poolers/samplers to top-level of `experimentalist`

There are several reasons for moving the poolers and samplers into files at the top level of `experimentalist`:

### `poolers` and `samplers` are now less relevant

The strict difference between `poolers` and `samplers` now has less meaning

Conceptually: Now we just have functions which sit in a pipeline or cycle and operate on the state. Some of them still act like poolers and samplers of old, but they look more similar.

Technically: In the old setup, poolers took no positional arguments by design, whereas samplers took one (conditions). Now all of the poolers and samplers might take any number of arguments, like the experiment_data, model, conditions from a step before, and we achieve that using only named arguments in the functions which act on `State`.

### It makes sense to gather related functions in one file

Some of the poolers and samplers are _obviously_ related – like the random pooler/sampler, falsification pooler/sampler. 

It is simpler to have these in the same file, so that you can share functions more easily between them.

This will also remove one level of complexity for contributors – they'll just be able to make an `experimentalist` with full freedom to use whatever they want. 

This will also remove one chunk of complexity from the cookiecutter template.

### Allows for backward compatibility whilst freeing up the good names we want

There are two things we really want to achieve here:
- no breaking changes in this release
- free up names like `grid_pool`, `random_sample`, `random_pool` for our new paradigm

The proposed method was to make the existing functions into [`singledispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch) functions which had different behavior depending on what the inputs were. However, this relies on there always being at least one positional argument to the function, an assumption which is broken for the old poolers which had no positional arguments by design.

An attempt was made to remedy this, but it was super complicated and didn't work because of the positional arguments thing. 

Both in this case, and in future cases where we'll want to update all of the existing contributions to use the new setup, it's just much easier to start over in a new file than to try to make it work in the same file.

### CODE REVIEW

File: docs/experimentalists/pooler/random/index.md
Comment (benwandrew) [inline docs/experimentalists/pooler/random/index.md:29]: should it be `num_samples` here instead of `n`?

on a related note, this prompted me to look through the `random_.py` file; why is it not also `num_samples=num_samples` in [line 339](https://github.com/AutoResearch/autora-core/blob/221302d5a4df65ba7c20a7b24fd87e514bf871fe/src/autora/experimentalist/random_.py#L339)

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:31]:
```suggestion
        s: a State object
        num_samples: the number of conditions to produce
        random_state: the seed value for the random number generator
        replace: if True, allow repeated values
```

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:32]:
```suggestion
```

**Review by benwandrew — CHANGES_REQUESTED** [review]

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:96]: does having this bit up on the previous line just make it too long for our linter?

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:170]: same question as above re: floating `x`

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:178]:
```suggestion
        ... With one independent variable "x", and a value_range we get a sample of the range back when running the experimentalist:
```

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:321]:
```suggestion
    Returns: a Result object with a field `conditions` that is constituted by a DataFrame of the sampled conditions
```

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:329]: same re: floating `x`

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:339]:
```suggestion
            conditions, random_state=random_state, num_samples=num_samples, replace=replace
```

**Review by benwandrew — CHANGES_REQUESTED** [review]
so fun working through this; i now (at least somewhat) understand `singledispatch`!

also, i think moving everything to top-level `experimentalist` directory and condensing related functions into a single file makes a lot of sense. complexity be gone!

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:34]: i've noticed that occasionally certain entries are left blank in docstrings (not just here, but generally); is this an intentional/reasoned in certain circumstances?

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [inline src/autora/experimentalist/random_.py:317]: same question as before; is there a reason for leaving these entries blank? seems like this one might just be an oversight, but less sure about `Returns` above

File: docs/experimentalists/pooler/grid/quickstart.md
Comment (musslick) [inline docs/experimentalists/pooler/grid/quickstart.md:13]: Why is it `autora.experimentalist.grid_` and not just `autora.experimentalist.grid`?

File: docs/experimentalists/pooler/random/index.md
Comment (musslick) [inline docs/experimentalists/pooler/random/index.md:27]: Same here. What is the reasoning for using `_` in this case, and not in other cases?

File: docs/experimentalists/pooler/random/index.md
Comment (musslick) [reply docs/experimentalists/pooler/random/index.md:29]: Yes, great catch, Ben! It should be `num_sampled`; not `n`.

File: src/autora/experimentalist/grid_.py
Comment (musslick) [inline src/autora/experimentalist/grid_.py:39]: Hmm.. I'm not sure how I feel about this output. I think it is much more intuitive if the pooler would return a numpy array or pandas data frame.

File: src/autora/experimentalist/random_.py
Comment (musslick) [inline src/autora/experimentalist/random_.py:261]: Hmm... So the random function returns a data frame, but the grid function returns some kind of iterable? Or am I missing something? In any case, this should be consistent.

**Review by musslick — CHANGES_REQUESTED** [review]
Looks great, just got a couple minor comments/clarifications!

File: docs/experimentalists/pooler/random/index.md
Comment (hollandjg) [reply docs/experimentalists/pooler/random/index.md:29]: Yes!

File: docs/experimentalists/pooler/grid/quickstart.md
Comment (hollandjg) [reply docs/experimentalists/pooler/grid/quickstart.md:13]: The problem here is tricky. When you run `pytest`, it imports any submodules using their unqualified name – it uses `grid` rather than `autora.experimentalist.grid`; `grid` rather than `autora.experimentalist pooler.grid`. 

Unfortunately, this means there are two things with the same name, and that throws an error.

For backwards compatibility, where we didn't want to break the old imports from `autora.experimentalist pooler.grid`, we want to leave that old module where it is. We have to have a different name, and the recommended way to do that is to add an underscore on the end.

File: docs/experimentalists/pooler/random/index.md
Comment (hollandjg) [reply docs/experimentalists/pooler/random/index.md:27]: `random` is a different case – that's the name of a python built-in module (as in `from random import seed, choice`). We shouldn't have a thing with the same name as a built-in (qualified or unqualified) because that will lead to a lot of confusion, so we need a different name for that.

File: src/autora/experimentalist/random_.py
Comment (hollandjg) [reply src/autora/experimentalist/random_.py:36]: no, pure failure to notice that they were missing. I'll update those. Thanks!

File: src/autora/experimentalist/random_.py
Comment (hollandjg) [reply src/autora/experimentalist/random_.py:98]: No, previously the function name was much longer. The formatting is fixed in the next update.

File: src/autora/experimentalist/random_.py
Comment (hollandjg) [reply src/autora/experimentalist/random_.py:263]: You're right, makes no sense! grid is fixed in the next update.

File: src/autora/experimentalist/random_.py
Comment (hollandjg) [reply src/autora/experimentalist/random_.py]: same again – failure to notice. Updated now.

**Comment by hollandjg** [comment]
> I think, we should talk about this in a meeting before landing on this design.

I'm totally up for that. It's pretty easy to change the names and the conventions now, but as soon as we've merged this it gets more difficult. @younesStrittmatter and @benwandrew and @musslick – when's a good time to chat about this?

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [reply src/autora/experimentalist/random_.py]: yes, that's better!

File: src/autora/experimentalist/random_.py
Comment (benwandrew) [reply src/autora/experimentalist/random_.py:342]: ah, ok. thanks for correcting me!

**Comment by younesStrittmatter** [comment]
I think it would be great to have a meeting as soon as possible. I would have time today, but I don't know about others. I guess if we can not find time before, we can also talk about this tomorrow in one of the meetings.

File: docs/experimentalists/pooler/random/index.md
Comment (benwandrew) [reply docs/experimentalists/pooler/random/index.md:27]: `chance`? `stochastic`? 

the latter is probably already too loaded as a term with a particular technical meaning...?

**Comment by benwandrew** [comment]
> > I think, we should talk about this in a meeting before landing on this design.
> 
> I'm totally up for that. It's pretty easy to change the names and the conventions now, but as soon as we've merged this it gets more difficult. @younesStrittmatter and @benwandrew and @musslick – when's a good time to chat about this?

i'm flexible today after 1:30 PM EST and tomorrow after 10:30 AM EST

**Comment by benwandrew** [comment]
> I think, we should talk about this in a meeting before landing on this design. From a user perspective, I somewhat like the single-dispatch thing, but I have some concerns:
> 
> * From a user perspective: It might not be clear, what is going on in the background. Why the function is working on some states, but not on others. What the inputs and outputs have to look like... (minor concerns)
> * From a contributor perspective: If this is a supposed to be a design template for other experimentalist, I have concerns about the additional complication a contributor faces when they have to use decorators. That might shy away people from contributing to autora.
> * Design problem: As soon as there are more than one state variable as input, the singledispatch approach doesn't work anymore, I would suspect. Then we need some other decorator (multidispatch, double single-dispatch?).

are decorators really not enough of basic building block for researchers that we can reasonably expect to use/contribute? genuinely asking, but seems like the concept is introduced fairly early on when learning Python. concerning multiple State inputs, if singledispatch is insufficient, looks like there are at least [one](https://github.com/mrocklin/multipledispatch/) multiple dispatch solution out there...

**Comment by hollandjg** [comment]
Action items after our discussion today (@younesStrittmatter @benwandrew @musslick):
- [x] Drop the single dispatch
- [ ] Make the 'base' function which has arguments like `def f(conditions, experiment_data, ...)` the one with the short name, and make the wrapped version which takes the state `def f_x(state, **kwargs)` the one with the longer name. 
- [ ] Decide on a convention for modifying the name, e.g. `f` -> {`f_wf` or `f_s` or `f_state` or `f_task` or `f_t` or `s_f`}, ask the group.

**Comment by hollandjg** [comment]
### Options for function naming (in no particular order, with best-effort attributions)

Overall solutions:
- submodules @chadcwilliams 
  - `pool.on_state`, `pool.raw` @musslick 
  - `pool.run`, `pool.run_on_state`
  - `random.sample_.on_state`, `random.sample_.raw`

Longer names for the wrapped function:
- `def s_random_pool(...` prefix
- `def random_pool_task(` @musslick 
- `def random_pool_wf(`
- `def random_pool_s(`
- `def random_pool_state(`
- `def random_pool_autora(` -> works on whatever the `Standard` is @TheLemonPig 
- `def random_pool_on_state(` @ioanam25

Longer names for the raw function:
- `def random_pool_raw(` @TheLemonPig

**Comment by hollandjg** [comment]
Some ideas from CCV:

- Make Delta the primary thing. (Each of these could be implemented using `__radd__`):
```python
from autora.experimentalist import grid_, random_
import autora.experiment_runner.synthetic.weber as experiment
import autora.theorist.darts as model
s_4 = s_0 + grid_.pool() + random_.sample() + experiment.observe() + model.fit()
```

- Make delta the primary thing, but make it explicit about the inputs
```python
from autora.experimentalist import grid_, random_
import autora.experiment_runner.synthetic.weber as experiment
from autora.theorist.darts import DARTSRegressor

model = wrap_regressor(DARTSRegressor())  # wrapper for .fit()

s_1 = s_0 + grid_.pool(s_0)
s_2 = s_1 + random_.sample(s_1)
s_3 = s_2 + experiment.observe(s_2)
s_4 = s_3 + model.fit(s_3)
```

- Each task uses the same generic verbs:
```python
import autora.experimentalist.grid_ as pool
import autora.experimentalist.random_.sample as sample
import autora.experiment_runner.synthetic.weber as experiment
import autora.theorist.darts as model

s_4 = model.run(experiment.run(sample.run(pool.run(s_0))))
```

- Each task uses some conventional verbs which describe what it's doing (sort of):
```python
from autora.experimentalist.grid_ import pool
from autora.experimentalist.random_ import sample  # (could also `import pool`)
from autora.experiment_runner.synthetic.weber import observe
from autora.theorist.darts import model

s_4 = model(observe(sample(pool(s_0))))
```

- Each task uses more verbose verbs:
```python
from autora.experimentalist.grid_ import generate_conditions
from autora.experimentalist.random_ import sample_conditions
from autora.experiment_runner.synthetic.weber import make_observations
from autora.theorist.darts import fit_model

s_4 = fit_model(make_observations(sample_conditions(generate_conditions(s_0))))
```

Contributors:
@digicosmos86
@tdivoll 
@gtdang 
@mirestrepo

**Comment by hollandjg** [comment]
- Make State the primary thing, but make it explicit about the inputs
```python
from autora.experimentalist import grid_, random_
import autora.experiment_runner.synthetic.weber as weber
import autora.theorist.darts as darts

s_1 = grid_.pool(s_0)
s_2 = random_.sample(s_1)
s_3 = weber.observe(s_2)
s_4 = darts.fit(s_3)
```

**Comment by hollandjg** [comment]
We came to the conclusion that:
- having the user wrap everything is more consistent
- having some standard wrappers might be nice
- exposing the `base` function as the standard name gives us the least work now (don't have to re-wrap everything) and the most flexibility for the future
- Users can always wrap things in different ways – using the wrapper functions with some function name mappings, or using a standard `def my_wrapper(s: State...)` or some other way we've not thought about yet.

**Comment by hollandjg** [comment]
- [ ] throw an error if a Delta field doesn't match the State

**Comment by hollandjg** [comment]
- `wrap_to_use_state` feels too verbose – what does it actually do?

**Comment by younesStrittmatter** [comment]
I think there is a misunderstanding here: If a function returns a thing that is not a field object. The thing doesn't get added to the state. Instead, the state just doesn't get changed. I am not sure, if there is any case, in which this is a desired outcome. (Maybe if a function returns two things and only one of them is used. But then the wrapper could also just not return the thing)

**Comment by younesStrittmatter** [comment]
As far as I understand it, one can not add fields to the state dynamically and has to extend the class instead. 
If a function is called that is defined on a non-existend field input, it will throw an error. 
If a function is called that is defined on a non-existend field output nothing will happen (the field doesn't get added, the state doesn't change)

**Comment by benwandrew** [comment]
> > * `wrap_to_use_state` feels too verbose – what does it actually do?
> 
> Alternatives:
> 
> `wrap_to_autora` `wrap_to_workflow` `wrap_to_task`

if there is still concern about "state" being a heavily overloaded CS term, any of these should alleviate that. if indeed the new state/delta setup becomes the mainstay for `autora`, then `wrap_to_autora` or `wrap_for_autora` seems nice...

**Comment by younesStrittmatter** [comment]
In my opinion, we should throw an error when trying to use fields that are not in the state (both input and output). And also customize the error that is thrown to point to the wrapper.  
I would also advocate for making the mapping from the input arguments explicit as well (instead of relying on the naming of the underlying functions).
Here is the why: This is an error a user gets when trying to use a wrapped function, where the underlying function has the wrong keyword-argument name. (In this case `test` instead of `conditions`). `experiment_runner` in this case, is the function, that the wrapper has returned and that should operate on the state.
```
experiment_runner() missing 1 required positional argument: 'test'
```
There is no way a user reads this error message and comes to the conclusion that the function they used in the wrapper had a wrong keyword argument. There is no mention of state/fields/ wrapping or anything like this.

**Comment by younesStrittmatter** [comment]
Maybe I would go even further and make the state itself and argument for the wrapper, so error checking can be done in the wrapper, instead of when running the function. This way it is clear, that the wrap_to_state function, wraps the function to a specific state.

**Comment by hollandjg** [comment]
- [ ] Check that the sklearn regressors can use data frames across the board before locking this in

**Comment by hollandjg** [comment]
Hi all, esp @musslick @younesStrittmatter @benwandrew @blinodelka @chadcwilliams @benwandrew 
I've updated this PR to include the core changes from the mini-hackathon. I think this might be ready to merge, so please let me know if there's something broken.

If anything is missing, I suggest we make that a new PR, because this one is pretty large and already does a lot of stuff. 

- I didn't include the `align_dataframe_to_ivs` function from https://github.com/AutoResearch/autora-core/blob/mini-hackaton-before-refactor-state/src/autora/utils/conversion.py because I don't think this should be part of the `on_state` function. 
  - The functions in `autora.state.delta` are by design completely agnostic about the contents of the state objects, other than assuming that they can be treated as `State` objects and are fundamentally `dataclasses`. That means that they don't do any transformations on the data beyond extending DataFrames and renaming things according to the alias rules. 
  - I think it's best to include the `align_dataframe_to_ivs` in the places where you need it in your experimentalist, for instance, or to define a new function wrapper which you can use if you need it. 
  - Remember that if you're using data frames you can just pass a list of column names and you've already got exactly those columns in that order, and I think that's probably more understandable to the end-user.

**Comment by hollandjg** [comment]
> So far, we have been using it in both experimentalists and synthetic models (autora). Any other place where we can put it in the core?

I can't think of any.

File: docs/experimentalists/pooler/random/index.md
Comment (hollandjg) [reply docs/experimentalists/pooler/random/index.md]: I've updated the GitHub Actions test runner to use `--import-mode importlib` so that we don't have this error any more. The module name is now `autora.experimentalist.random`

**Comment by hollandjg** [comment]
Hi all, esp. @musslick @benwandrew and @younesStrittmatter – I'd appreciate your re-review of this PR! Please let me know if there's anything else needing changing.

File: src/autora/state.py
Comment (musslick) [inline src/autora/state.py:1367]:
```suggestion
```

File: src/autora/state.py
Comment (musslick) [inline src/autora/state.py:1280]:
```suggestion
def estimator_on_state(estimator: BaseEstimator) -> StateFunction:
```

File: src/autora/state.py
Comment (musslick) [inline src/autora/state.py:1322]:
```suggestion
def experiment_runner_on_state(f: Callable[[X], Y]) -> StateFunction:
```

File: src/autora/state.py
Comment (musslick) [inline src/autora/state.py:1338]:
```suggestion
        >>> experiment_runner_on_state(x_to_y_fn)(s).experiment_data
```

File: src/autora/state.py
Comment (musslick) [inline src/autora/state.py:1368]:
```suggestion
experiment_runner_on_state(f: Callable[[X], XY]) -> StateFunction:
```

Comment (musslick) [inline src/autora/state.py:1368]:
```suggestion
def experiment_runner_on_state(f: Callable[[X], XY]) -> StateFunction:
```

**Review by musslick — CHANGES_REQUESTED** [review]
Looks great but docstrings need to be fixed

**Review by musslick — APPROVED** [review]
Looks great!

**Pull request merged**
